cpp/src/join/distinct_hash_join.cu

```diff
@@ -205,27 +205,24 @@ CUDF_KERNEL void distinct_join_probe_kernel(Iter iter,
     cudf::size_type buffer_size = 0;
 
     while (idx - block.thread_rank() < n) {  // the whole thread block falls into the same iteration
-      cudf::size_type thread_count{0};
-      cudf::size_type build_idx{0};
-      if (idx < n) {
-        auto const found = hash_table.find(*(iter + idx));
-        thread_count     = found != hash_table.end();
-        build_idx        = static_cast<cudf::size_type>(found->second);
-      }
+      auto const found     = idx < n ? hash_table.find(*(iter + idx)) : hash_table.end();
+      auto const has_match = found != hash_table.end();
 
       // Use a whole-block scan to calculate the output location
       cudf::size_type offset;
       cudf::size_type block_count;
-      block_scan(block_scan_temp_storage).ExclusiveSum(thread_count, offset, block_count);
+      block_scan(block_scan_temp_storage)
+        .ExclusiveSum(static_cast<cudf::size_type>(has_match), offset, block_count);
 
       if (buffer_size + block_count > buffer_capacity) {
         flush_buffer(block, buffer_size, buffer, counter, build_indices, probe_indices);
         block.sync();
         buffer_size = 0;
       }
 
-      if (thread_count == 1) {
-        buffer[buffer_size + offset] = cuco::pair{build_idx, static_cast<cudf::size_type>(idx)};
+      if (has_match) {
+        buffer[buffer_size + offset] = cuco::pair{static_cast<cudf::size_type>(found->second),
+                                                  static_cast<cudf::size_type>(idx)};
       }
       buffer_size += block_count;
       block.sync();
```

Comment:
*question*: I can't comment on the relevant line, but do you need a `block.sync()` on line 235, after the final `flush_buffer`?

Reply:
The `block.sync()` after `flush_buffer` is mainly to guarantee that the tile counter is properly reset after writing. The final `flush_buffer` doesn't need `block.sync()` since the counter won't be used anymore. I could miss something obvious, what's your consideration regarding the `block.sync()` there?

Reply:
Thanks, nothing to do. I was just wondering if this kernel might be called in a loop on device.